README.md
![Logo](static/img/wcac_logo.png)

Welt Content API Client
=======================

Build
-----

    ./sbt clean coverage test
    ./sbt coverageReport
    ./sbt coverageAggregate

Local Development (with an ApiClient-Project)
---------------------------------------------

There is no need to do a full dev cycle (publish to bintray and bump the version in the other project)
each time something changes. For this use case you can benefit from _SNAPSHOT_ releases. Set the version to
`version := "0.2.1-SNAPSHOT",` and then do a

    ./sbt publishLocal

in the other project, enable resolving vom local caches by adding a resolver `Resolver.mavenLocal` and also
switch to the same _SNAPSHOT_ version.

Publish
-------
Publishing Versions is done by commiting to master. CI [Jenkins](https://jenkins.up.welt.de/job/funkotron/job/WeltContentApiClient/job/WeltContentApiClientPipeline/) will take care of the rest for you.

(!) Bump the version in file `build.sbt`, otherwise you will get an error when publishing a version that already exists.

We are using the [bintray sbt plugin](https://github.com/softprops/bintray-sbt).

Deploying from local
--------------------
You will need a [bintray](https://bintray.com/) account. Log in through sbt:

    ./activator bintrayChangeCredentials

Then you can publish new releases with:

    ./sbt publish

Service Authentication
----------------------

![IAM](static/img/wcapi_auth.png)

- when using the _wcapi_ on your machine (_funkotron_) it is required to:
    - Have a valid AWS account, including `credentials` (`access key` and `secret access key` -> ask your admin).
    The easiest way is to [install the AWS CLI][aws_cli] and then execute `aws configure`.
    - Have the `frontend` profile configured as follows:

~/.aws/config
```
[profile frontend]
region = eu-west-1
role_arn = arn:aws:iam::933782373565:role/dev_restricted_access
source_profile = default
```

~/.aws/credentials
```
[default]
aws_access_key_id = XXX
aws_secret_access_key = XXX
```

- the _AWS services_ (e.g. S3, SSM and all the future services) are required to use the `Credentials` provided by the _wcapi_

```scala
import com.amazonaws.auth.AWSCredentialsProvider
import de.welt.contentapi.core.client.services.configuration.ApiConfiguration
import scala.util.Try

val credentials: Try[AWSCredentialsProvider] = ApiConfiguration.aws.credentials
```

- Configurations are automatically loaded from SSM
- You should set the local mode to `dev` in the `application.conf` like:

```hacon
content_api {
    mode: "dev"
    # this is usually overriden in the ecs task-definition.json
    mode = ${?MODE}
}
```

- local overrides can be provided by writing a file to `~/.welt/${module}/frontend-overrides.conf`, e.g.:

```hocon
api {
  cigger {
    host: ""
    endpoint: ""
    method: ""
    basic_username: ""
    basic_password: ""
    api_key: ""
    circuit_breaker {
      enabled: false
      max_failures: 5
      call_timeout: 5
      reset_timeout: 10
      exponential_backoff: 2
    }
  }
}
s3 {
  raw_tree {
    bucket: "not-the-prod-bucket"
    file: "/not/the/prod/file.json"
  }
}
```

- client projects should generate a standardized `version.conf`, please add this to your `build.sbt`:

```sbt
resourceGenerators in Compile += Def.task {
  val file = (resourceDirectory in Compile).value / "version.conf"
  IO.write(file,
    s"""build_info {
       |    version  = "${version.value}"
       |    revision = "${git.gitHeadCommit.value.getOrElse("cannot_detect_git_revision")}"
       |    module   = "${packageName.value}"
       |    date     = "${if (version.value.contains("local")) "prevent reload loops in local dev" else java.time.Instant.now()}"
       |}""".stripMargin)
  Seq(file)
}.taskValue
```

[aws_cli]: https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html


Changelog.md
# Changelog

## 5.3 (2019-10-30)

### Changes
- [RAW] removed mandatory field `adIndicator` as part of `RawArticlePromotion`

## 5.2 (2019-10-25)

### Changes
- [RAW] added mandatory field `adIndicator` as part of `RawArticlePromotion`

## 5.1 (2019-10-23)

### Changes
- [CORE] Transform 401 to 502 in AbstractService to preserve Akamai and Varnish cache in case of misconfiguration

## 5.0 (2019-10-08)

### Changes
- [RAW] RawReads migrate old Header and Sponsoring to Sitebuilding
- [API] RawToApiConverter now does intelligent inheritance within all the Sitebuilding fields

## 4.7 (2019-09-23)

### Changes
- [RAW] added JSON serializer `allChildrenWrites` as part of `PartialRawChannelWrites`

## 4.6 (2019-09-18)

### Changes
- [RAW] added optional field `description` as part of `RawChannelStageConfiguredId`

## 4.5 (2019-08-15)
- changed the way configs are passed to the `AbstractService` to prevent lazy/null exceptions 
- Play 2.7.3
- Scala 2.12.9 and Scala 2.13.0

## 4.4 (2019-08-13)
- Added API method to find channels that have a particular commercial stage

## 4.3 (2019-07-30)
- add new field to `RawChannelConfiguration`: `articlePromotions: Option[Seq[RawArticlePromotion]]`
- make some Api fields mutable to be able to modify them in Cigger to migrate Content to Inline Elements for inline promotions

## 4.2 (2019-04-25)

- play 2.7.1
- using `com.kenshoo.metrics-play` instead of fork `de.welt.metrics-play`
- catching `IllegalArgumentException` when register `circuit_breaker` metric (only logging now)
- more version bumps
- clean up code

## 4.1 (2019-04-11)

- changed config for `CapiExecutionContext`

```hacon
// before
contexts.capi {
    fork-join-executor {
        parallelism-factor = 25
        parallelism-min = 25
        parallelism-max = 100
    }
}
// after
content_api {
    akka {
        fork-join-executor {
            parallelism-factor = 25
            parallelism-min = 25
            parallelism-max = 100
        }
    }
}
```

- change path for `frontend-overrides.conf` (incl. app name)
- allowed `prod` and `development` as a valid stage/mode name

## 4.0 (2019-03-01)

- Merged the `client` libs with the `core` libs reducing the lib and module count to `core`, `pressed` and `raw`
- removed the menu stuff since it is mainly used in cmcf
- Rewrote the `configuration` to use the _AWS Simple Systems Manager_ instead of Play's config mgmt
- Removed the `Env` (Live,Preview) since it has not been used anywhere 

## 3.4 (2019-02-26)
- Added API to find channels that have a particular stage configured from god mode

## 3.3 (2019-02-04)

### Changes
- Updated to Playframework 2.7.0

## 3.2 (2018-11-21)

### Changes
- [RAW] changed default `RawChannelSiteBuilding` configuration from *empty Optional* to *None*
- implemented master channel inheritance for `RawChannelSiteBuilding`

## 3.1 (2018-11-05)

### Changes
- [RAW] added optional field `logo` as part of `RawChannelStageCustomModule`

## 3.0 (2018-10-12)

*Changes*

- (breaking) changed signature of the `AbstractSerivce` to:
  - Allow other `methods` than get by setting the `method: "POST"` in the ServiceConfig
  - Allow other content types than `JSON` by changing the signature of the `validate` function 
  (see changes -> `ContentService` for an example implementation)
  - Made `basic auth credentials`/`Api key` optional: Its now possible to omit credentials 

## 2.3 (2018-10-04)

*Changes*

- added `RawChannelSiteBuilding` for configuring channel header, footer and sponsoring.

*Deprecations* 

- Deprecated `RawChannelHeader` and `RawSponsoringConfig`

## 2.2 (2018-09-18)

- `MenuService`: removed auto sync cron job. It's now part of the consumer.
- `MenuService`: add `refresh()` for sync data from S3.
- `MenuService`: add new `Configuration` value `mode` to override the S3 env folder. This is for testing prod data on localhost.

## 2.1 (2018-08-29)

All WC/AC models must include a prefix for better identification in code reviews.
Bring your own models (BYOM) => consume the api + api models but create new models for your application logic / templates.

- renamed case classes: `Menu` -> `ApiMenu`, `MenuMetadata` -> `ApiMenuMetadata` and `MenuLink` -> `ApiMenuLink`
- removed `MenuReference` (replaced by `ApiReference`)
- refactoring of `MenuService` and `AdminMenuService`

## 2.0 (2018-08-09)

- added new project called `menu`
- added `MenuService` for retrieving menu data from S3
- added case classes `Menu`, `MenuMetadata`, `MenuLink` and `MenuReference`

#### Breaking Changes

- Moved `AdminMenuService` to `menu` project
- Removed case classes `RawMenu` and `RawMenuLink`

## 1.8 (2018-08-02)

- added `RawChannelStageConfiguredId` as a new main stage type
- this allows single ids to be configured on section pages without the need of Papyrus Curation
- example usecases: Advertorial Footers, oEmbeds (WM, Podcasts,...)


## 1.7 (2018-06-08)

- added `AdminMenuService` for storing menu data on S3
- added case classes `RawMenu` and `RawMenuLink`

## 1.6 (2018-06-04)

- Added optional sponsoring to `ApiStageConfiguration`

## 1.5 (2018-05-31)

*Changes*

- introduced new `TreeService` as an abstraction to the `RawTreeService` in the front end
- Migrated the `AuthorService` from _funkotron_ into the capi to allow using it within cigger (why: _author augmenting_)
- `RawTreeService` will not access s3 when app is started in `Mode.Test`
- Play Framework version bump to `2.6.15`

*Deprecations*

- Deprecated large parts of `ApiAuthor` in favor of augmenting authors json into the `related` array of the `ApiPressedContent`

## 1.4 (2018-03-28)

- Changed signature of _cigger_ to `ApiPressedContentResponse`

## 1.3 (2018-03-27)

- added `ApiPressedContentResponse`
- bumped to Play 2.6.12

## 1.3 (2018-03-20)

- add optional http headers in AbstractService

## 1.2 bugfix (2018-03-16)

- definition which Exceptions will cause the breaker to stay open or allow close. e.g.

project/MyBuild.scala
import bintray.BintrayPlugin.autoImport._
import com.scalapenos.sbt.prompt.SbtPrompt.autoImport._
import org.scalastyle.sbt.ScalastylePlugin._
import sbt.Keys._
import sbt._
import scoverage.ScoverageSbtPlugin.autoImport._


object MyBuild extends Build {

  val buildNumber = Option(System.getenv("BUILD_NUMBER")).getOrElse("local-SNAPSHOT")
  val forScala2_4 = Option(System.getenv("PLAY24")).exists(_.toBoolean)

  val playVersion = if (forScala2_4) "2.4.8" else "2.5.10"
  private val actualVersion: String = s"0.21.$buildNumber"

  scalaVersion := "2.11.8"

  def withTests(project: Project) = project % "test->test;compile->compile"

  val frontendCompilationSettings = Seq(
    organization := "de.welt",
    scalaVersion := "2.11.8",
    version := s"$actualVersion${if (forScala2_4) "_2.4.0" else ""}",

    licenses += ("MIT", url("http://opensource.org/licenses/MIT")),
    javacOptions ++= Seq("-source", "1.8", "-target", "1.8"),
    publishArtifact in Test := false,
    promptTheme := com.scalapenos.sbt.prompt.PromptThemes.ScalapenosTheme
  )

  val frontendDependencyManagementSettings = Seq(
    resolvers := Seq(
      //      Resolver.typesafeRepo("releases"),
      Resolver.jcenterRepo
    ),
    // https://www.typesafe.com/blog/improved-dependency-management-with-sbt-0137
    updateOptions := updateOptions.value.withCachedResolution(true)
  )

  val coreDependencySettings = Seq(
    libraryDependencies ++= Seq(
      "com.typesafe.play" %% "play" % playVersion % Provided,
      "com.typesafe.scala-logging" %% "scala-logging" % "3.5.0",
      "org.scalatestplus.play" %% "scalatestplus-play" % "1.5.1" % Test,
      "org.mockito" % "mockito-core" % "1.10.19" % Test,
      "com.typesafe.play" %% "play-cache" % playVersion % Provided,
      // Info:
      // Prevent compile warnings of: `Class javax.annotation.Nullable not found - continuing with a stub.`
      "com.google.code.findbugs" % "jsr305" % "3.0.1" % Compile
    )
  )
  val clientDependencySettings = Seq(
    libraryDependencies ++= Seq(
      "org.asynchttpclient" % "async-http-client" % "2.0.10",
      "ch.qos.logback" % "logback-classic" % "1.1.7",

      "com.amazonaws" % "aws-java-sdk-core" % "1.11.13",
      "com.amazonaws" % "aws-java-sdk-s3" % "1.11.13",

      "com.typesafe" % "config" % "1.3.0" % Provided,

      "com.typesafe.play" %% "play-ws" % playVersion % Provided,
      "com.typesafe.play" %% "play-cache" % playVersion % Provided,
      "com.typesafe.scala-logging" %% "scala-logging" % "3.5.0",

      "org.scalatestplus.play" %% "scalatestplus-play" % "1.5.1" % Test,
      "org.mockito" % "mockito-core" % "1.10.19" % Test
    ) ++ (
      if (forScala2_4) {
        Seq("com.kenshoo" %% "metrics-play" % "2.4.0_0.4.1")
      }
      else {
        Seq("de.threedimensions" %% "metrics-play" % "2.5.13")
      })
  )


  val bintraySettings = Seq(
    pomExtra :=
      <scm>
        <url>git@github.com:WeltN24/{name.value}.git</url>
        <connection>scm:git:git@github.com:WeltN24/{name.value}.git</connection>
      </scm>
        <developers>
          <developer>
            <id>thisismana</id>
            <name>Matthias Naber</name>
            <url>https://github.com/thisismana</url>
          </developer>
          <developer>
            <id>harryurban</id>
            <name>Harry Urban</name>
            <url>https://github.com/harryurban</url>
          </developer>
          <developer>
            <id>bobaaaaa</id>
            <name>Patrick Dahms</name>
            <url>https://github.com/bobaaaaa</url>
          </developer>
        </developers>,
    bintrayRepository := "welt-content-api-client",
    bintrayOrganization := Some("welt"),
    bintrayVcsUrl := Some("git@github.com:you/your-repo.git")
  )

  def codeStyleSettings = Seq(
    // scoverage
    coverageExcludedPackages := "<empty>;",
    //    coverageMinimum := 25, // we are not that good yet ;)
    coverageFailOnMinimum := true,

    // scalastyle
    scalastyleConfig := file("scalastyle-config.xml"),
    scalastyleFailOnError := true
  )

  def project(id: String) = Project(id, base = file(id))
    .settings(frontendCompilationSettings: _*)
    .settings(frontendDependencyManagementSettings: _*)
    .settings(bintraySettings: _*)
    .settings(codeStyleSettings: _*)


  // only in "welt-content-api-client"
  val utils = project("utils")
    .settings(
      name := "welt-content-api-utils"
    )
    .settings(coreDependencySettings: _*)

  val core = project("core")
    .settings(
      name := "welt-content-api-core"
    )
    .settings(coreDependencySettings: _*)

  val raw = project("raw")
    .settings(clientDependencySettings: _*)
    .settings(
      name := "welt-content-api-raw"
    )
    .dependsOn(withTests(utils)).aggregate(utils)
    .dependsOn(withTests(core)).aggregate(core)

  val pressed = project("pressed")
    .settings(
      name := "welt-content-api-pressed"
    )
    .settings(coreDependencySettings: _*)
    .dependsOn(withTests(core)).aggregate(core)

  val coreClient = project("core-client")
    .settings(
      name := "welt-content-api-core-client"
    )
    .settings(clientDependencySettings: _*)
    .dependsOn(withTests(utils)).aggregate(utils)
    .dependsOn(withTests(core)).aggregate(core)

  val rawClient = project("raw-client")
    .settings(
      name := "welt-content-api-raw-client"
    )
    .settings(clientDependencySettings: _*)
    .dependsOn(withTests(coreClient)).aggregate(coreClient)
    .dependsOn(withTests(raw)).aggregate(raw)

  val pressedClient = project("pressed-client")
    .settings(
      name := "welt-content-api-pressed-client"
    )
    .settings(clientDependencySettings: _*)
    .dependsOn(withTests(coreClient)).aggregate(coreClient)
    .dependsOn(withTests(pressed)).aggregate(pressed)
    .dependsOn(withTests(rawClient)).aggregate(rawClient)

  val main = Project("Root", base = file("."))
    .settings(
      name := "welt-content-api-root"
    )
    .settings(frontendCompilationSettings: _*)
    .settings(
      publish := {},
      bintrayUnpublish := {}
    )
    .aggregate(core, coreClient, raw, rawClient, pressed, pressedClient)

}
